File: Lab_8_2.py
def count_primes(n):
    # Хэрэв n нь 2-оос бага байвал 0-ыг буцаана
    if n < 2:
        return 0

    # Бүх тоонуудыг prime гэж үзэж эхэлнэ
    primes = [True] * (n + 1)
    primes[0], primes[1] = False, False  # 0 болон 1 нь prime биш

    # n-ын квадрат язгуур хүртэлх бүх тоонуудыг шалгана
    for i in range(2, int(n**0.5) + 1):
        if primes[i]:
            # Хэрэв prime байвал түүний бүх үржвэрүүдийг prime биш гэж тэмдэглэнэ
            for j in range(i * i, n + 1, i):
                primes[j] = False

    # prime массивын үнэн утгуудыг нийлбэрлэн prime тоонуудын тоог буцаана
    return sum(primes)

# Жишээ:
n = 18
print(count_primes(n))  # Гаралт: 7
# 2. Prime Number Calculation (Анхны тоо олох):
# (Time Complexity): O(n log log n), Сieve of Eratosthenes алгоритм ашигласан.
# (Space Complexity): O(n), учир нь prime массивыг ашиглаж байгаа.

File: Lab_8_2.java
import java.util.Arrays;

public class Lab_8_2 {
    public static int countPrimes(int n) {
        // Хэрэв n нь 2-оос бага байвал 0-ыг буцаана
        if (n < 2) {
            return 0;
        }

        // Бүх тоонуудыг prime гэж үзэж эхэлнэ
        boolean[] primes = new boolean[n + 1];
        Arrays.fill(primes, true);
        primes[0] = primes[1] = false;  // 0 болон 1 нь prime биш

        // n-ын квадрат язгуур хүртэлх бүх тоонуудыг шалгана
        for (int i = 2; i * i <= n; i++) {
            if (primes[i]) {
                // Хэрэв prime байвал түүний бүх үржвэрүүдийг prime биш гэж тэмдэглэнэ
                for (int j = i * i; j <= n; j += i) {
                    primes[j] = false;
                }
            }
        }

        // Prime үнэн утгуудын тоог нийлбэрлэн буцаана
        int count = 0;
        for (boolean prime : primes) {
            if (prime) count++;
        }
        return count;
    }

    public static void main(String[] args) {
        int n = 18;
        System.out.println(countPrimes(n));  // Гаралт: 7
    }
}